### PR TITLE
feat: undo-block tool

### PR DESCRIPTION
**Pull request by ppca**
undo-block tool to revert the current head of the chain to its previous block. Typically used when the state of current chain head is messed up, e.g. running a wrong binary like in the case of protocol upgrade or gas price param changes.

To use the tool:
1. stop the node
2. run command `./target/release/neard --home {path_to_config_directory} undo-block`
3. restart node

Assuming only the head of the chain was messed up, not a long fork, then after using the tool, the node should be able to catch up and produce block (if they are block producers).

We intentionally forbid users from reverting beyond final block so that nothing is messed up for flat storage and finality is not broken.

[Design doc for the tool](https://docs.google.com/document/d/1bqdGtMCNwyHDNGWmcsVCR23q1BdcK2tu6ab87Kd43r0/edit#heading=h.dzkwz45fxs82)

### CODE REVIEW

File: tools/reset-head/src/cli.rs
Comment (ppca) [inline tools/reset-head/src/cli.rs:64]: commenting this out is w/o removing block data

File: chain/chain/src/store.rs
Comment (ppca) [inline chain/chain/src/store.rs:2289]: @pugachAG @Longarithm flat state column changes are here, please help review

File: tools/undo-block/src/cli.rs
Comment (ppca) [inline tools/undo-block/src/cli.rs:31]: when I compile nearcore with `--features=protocol_feature_flat_state`, and when I run the undo-block tool, I get this error for this line: 
`thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: IO(Custom { kind: Other, error: "Invalid argument: Column families not opened: FlatStorageStatus, FlatStateDeltaMetadata, FlatStateChanges, FlatState" })', tools/undo-block/src/cli.rs:31:66
`
How to address this?

Comment (ppca) [inline tools/undo-block/src/cli.rs:31]: When I compile nearcore with `--features=protocol_feature_flat_state` and run localnet with the binary, then run the undo-block tool (which uses same binary), this line causes error:
``
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: IO(Custom { kind: Other, error: "Invalid argument: Column families not opened: FlatStorageStatus, FlatStateDeltaMetadata, FlatStateChanges, FlatState" })', tools/undo-block/src/cli.rs:31:66
``

File: chain/chain/src/store.rs
Comment (nikurt) [inline chain/chain/src/store.rs:2253]: I find `for the purpose` comments not helpful. I'd much prefer to describe what the function does. And maybe somebody in the future decides that this function is useful in more usecases.

File: chain/chain/src/store.rs
Comment (nikurt) [inline chain/chain/src/store.rs:2254]: Please verify that the block is the current head.

File: chain/chain/src/store.rs
Comment (nikurt) [inline chain/chain/src/store.rs:2278]: Seems to be true. How did you find this out? Should someone create an issue to remember to fix it?

File: neard/src/cli.rs
Comment (nikurt) [inline neard/src/cli.rs:128]: Needed?

File: neard/src/cli.rs
Comment (nikurt) [inline neard/src/cli.rs:247]: Please add a comment, because this comment will be displayed when running `./neard` without a sub-command.

File: tools/undo-block/src/cli.rs
Comment (nikurt) [inline tools/undo-block/src/cli.rs:23]: I wonder if you need cold config at all. As you're only dealing with the head, it can't possibly affect cold storage.

File: tools/undo-block/src/cli.rs
Comment (nikurt) [inline tools/undo-block/src/cli.rs:46]:
```suggestion
        tracing::info!(target: "neard", %current_head_hash, %prev_hash, "Trying to update head");
```

Same in other `tracing::info!` calls.

File: tools/undo-block/src/cli.rs
Comment (nikurt) [inline tools/undo-block/src/cli.rs:1]: This file isn't just the CLI, but also the business logic.
I suggest to make this file `lib.rs`.

File: tools/undo-block/src/cli.rs
Comment (ppca) [reply tools/undo-block/src/cli.rs:31]: This happens because I ran localnet with a binary built without flat storage enabled( no --features=protocol_feature_flat_state flag), and when i run the undo-block tool, I ran a binary built with flat storage enabled. The issue is gone when i start localnet and run un-do block with binary that enables flat storage.

File: tools/undo-block/src/lib.rs
Comment (nikurt) [inline tools/undo-block/src/lib.rs:47]: Why 2 separate lines?

suggestion
```
tracing::info!(target: "neard", ?prev_block_hash, ?current_head_hash, ?prev_block_height, ?current_head_height, "Trying to update head");
```

Comment (nikurt) [inline tools/undo-block/src/lib.rs:47]: Why 2 separate lines?

```
tracing::info!(target: "neard", ?prev_block_hash, ?current_head_hash, ?prev_block_height, ?current_head_height, "Trying to update head");
```

Same below

**Review by nikurt — APPROVED** [review]
I don't understand why the flat storage part isn't working.
The rest looks good to me.

**Review by Longarithm — CHANGES_REQUESTED** [review]
Thank you! The code makes sense and flat storage removals are clear, I'm not sure what doesn't work for it.
Left some comments. I think it also deserves some test. Because it could be hard to implement, we can test the following scenario:
* create chain
* process some blocks
* undo last block H
* create new chain instance, check that chain metadata (tip?) is correct
* process some other block on height H, check that it works

**Comment by ppca** [comment]
@Longarithm 
we can test the following scenario:

create chain
process some blocks
undo last block H
create new chain instance, check that chain metadata (tip?) is correct
process some other block on height H, check that it works

--- 
do you mean testing using the tool for the above scenarios?
what does create chain mean here?
I've tested the following in localnet: keep sending transactions to localnet, stop one node, undo-block, restart. The node is able to catch up and produce blocks. I've tested stopping in the middle of an epoch, at the first block of an epoch and at the last block of an epoch. 
I've also tested the scenario of starting localnet and sending transaction, one node restarts with wrong gas cost, which creates an invalid state transition, and the node keeps doing header sync but not able to body sync. Then stop node, undo-block, restart with correct binary, and the node is able to catch up and produce blocks. I've tested stopping in the middle of an epoch, at the first block of an epoch and at the last block of an epoch. 
Do you mean some other scenarios to test?

File: tools/undo-block/src/lib.rs
Comment (ppca) [reply tools/undo-block/src/lib.rs:51]: this tool only removes head block and set head to head's prev block. This tool is unable to undo a random block in the middle of a chain.

**Comment by Longarithm** [comment]
I mean integration test for CI, sorry for confusion. The logic is not trivial, and it could be great to have undo block scenarios covered, so that we don't break it by chain/runtime/other changes. Especially if we want to use this code during outages.

Though integration tests are hard to write. I also don't see simple guidance for it, perhaps that's a good action item for core team. Let me give some introduction:

Integration tests live in `integration-tests` package. Common place for testing infra for processing blocks is `integration-tests/src/tests/client/process_blocks.rs`. The simplest setup would be
```
let mut genesis = Genesis::test(vec!["test0".parse().unwrap(), "test1".parse().unwrap()], 1);
let chain_genesis = ChainGenesis::new(&genesis);
let store = create_test_store();
let runtimes: Vec<Arc<dyn RuntimeWithEpochManagerAdapter>> =
        vec![nearcore::NightshadeRuntime::test(Path::new("../../../.."), store.clone(), genesis)];
let mut env = TestEnv::builder(chain_genesis).runtime_adapters(runtimes).build()
```
This one creates `TestEnv` with a single client and all structs necessary to test block production/processing logic.
First, we can create several blocks on chain:
```
let mut blocks = vec![];
for i in 1..5 {
    let block = env.clients[0].produce_block(i).unwrap().unwrap();
    blocks.push(block.clone());
    env.process_block(0, block, Provenance::PRODUCED);
}
```
Then, extract store update and undo block:
```
let mut chain_store_update = env.clients[0].chain.store().store_update();
chain_store_update.clear_block_data_undo_block(*blocks.last().unwrap().hash());
```
Then drop existing env, create new one **using old store** and try process new blocks on it. You can test your scenarios in that framework as well!
```
let runtimes: Vec<Arc<dyn RuntimeWithEpochManagerAdapter>> =
        vec![nearcore::NightshadeRuntime::test(Path::new("../../../.."), store, genesis)];
let mut env = TestEnv::builder(chain_genesis).runtime_adapters(runtimes).build()
for i in 3..7 {
    let block = env.clients[0].produce_block(i).unwrap().unwrap();
    env.process_block(0, block, Provenance::PRODUCED);
}
```

Example of test which kinda does it is `test_flat_storage_creation_sanity` we wrote recently.
Does it make sense?

File: chain/chain/src/store.rs
Comment (Longarithm) [inline chain/chain/src/store.rs:1960]:
```suggestion
   fn clear_chunk_data_and_headers_at_height(
```
because it is not called from outside, and it looks like impl detail of undoing block

**Comment by ppca** [comment]
> I mean integration test for CI, sorry for confusion. 

Makes sense to add a test for this in CI. I'll work on adding some for this tool.

File: tools/undo-block/src/lib.rs
Comment (Longarithm) [reply tools/undo-block/src/lib.rs:51]: Oh, sure! But then it might be cleaner to rename `clear_block_data_undo_block` to `clear_chain_head` and retrieve the chain head within the method instead of hash comparison, as it seems to be specifically designed for that purpose.

File: chain/chain/src/store.rs
Comment (Longarithm) [reply chain/chain/src/store.rs:2291]: Yeah, we don't need anything else. `home_dir` is used to read genesis, `store` - at least for reading state data, `config`... for tweaking execution parameters, caching, etc.

File: chain/chain/src/store.rs
Comment (ppca) [reply chain/chain/src/store.rs:2255]: I will refactor code a bit. this fn now changed to `pub fn clear_head_block_data(&mut self)`, and I separate out a `pub fn undo_block(chain_store)` which does the main changes to rocksdb. Will test undo_block in integration tests.

File: integration-tests/src/tests/client/undo_block.rs
Comment (Longarithm) [inline integration-tests/src/tests/client/undo_block.rs:25]:
```suggestion
    let runtime: Arc<dyn RuntimeWithEpochManagerAdapter> = nearcore::NightshadeRuntime::test(Path::new("../../../.."), store, genesis);
    (
        TestEnv::builder(chain_genesis).runtime_adapters(vec![runtime.clone()]).build(),
        runtime,
    )
```
nit

File: integration-tests/src/tests/client/undo_block.rs
Comment (Longarithm) [inline integration-tests/src/tests/client/undo_block.rs:41]: debug leftover?

File: integration-tests/src/tests/client/undo_block.rs
Comment (Longarithm) [inline integration-tests/src/tests/client/undo_block.rs:39]:
```suggestion
    for i in 1..=stop_height {
```

File: integration-tests/src/tests/client/undo_block.rs
Comment (Longarithm) [inline integration-tests/src/tests/client/undo_block.rs:28]: Deserves short comment, like "Checks that Near client can successfully undo block on given height and process it again after restart"

**Review by Longarithm — APPROVED** [review]
Thank you! 
Left some more comments but approving to unblock. 
The complexity of change makes me think that our current store codebase is poor :( Looking forward to further improvements, perhaps in scope of #8153.

File: chain/chain/src/store.rs
Comment (ppca) [reply chain/chain/src/store.rs:2226]: Ah I meant keyed by

**Comment by bowenwang1996** [comment]
@ppca what prevents us from merging this PR?

**Comment by ppca** [comment]
> @ppca what prevents us from merging this PR?

Was testing this on a testnet validator. The test looks good this morning. Will merge it today.

**Pull request merged**
